File: 250. Count Univalue Subtrees/countUnivalSubtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class countUnivalSubtrees {
    int res = 0;
    public int solution(TreeNode root) {
        if (root == null) {
            return 0;
        }
        dfs(root);
        return res;
    }

    public boolean dfs(TreeNode root) {
        boolean curr = true;
        if (root == null) {
            return true;
        }
        boolean left = true;
        boolean right = true;
        if (root.left != null) {
            left = dfs(root.left);
            if (root.left.val != root.val) {
                curr = false;
            }
        }
        if (root.right != null) {
            right = dfs(root.right);
            if (root.right.val != root.val) {
                curr = false;
            }
        }
        if (curr && left && right) {
            res++;
        }
        return curr && left && right;
    }
}

File: 250. Count Univalue Subtrees/countUnivalSubtrees.py
class countUnivalSubtrees:
    def solution(self, root):
        self.count = 0
        self.is_valid_part(root, 0)
        return self.count

    def is_valid_part(self, node, val):
        if node is None:
            return True
        if not all([self.is_valid_part(node.left, node.val),
                    self.is_valid_part(node.right, node.val)]):
            return False
        self.count += 1
        return node.val == val